refactor: simplify source code && update `webpack-defaults v1.1.0...1.5.0`

mattlewis92 opened the pull request:
It looks like travis / appveyor aren't enabled on this repo yet so the tests won't run, @d3viant0ne would you mind enabling these when you have a sec please?

I think this should be enough for an RC release. Any feedback is welcome! 

p.s. I removed all the sourcemap stuff from the codebase as it looks like the loader API in webpack 2+ changed so you can't get the source map from the previous loader and webpack will auto merge sourcemaps for you. Previously sourcemaps were always generated by default due to a bug in the source so I've left that on by default to preserve backwards compatibility (99% of the time users are going to want to produce sourcemaps, otherwise it will be impossible to debug their code)

hulkish commented:
@mattlewis92 Can you also include compilation  warnings/errors in your snapshot tests? example: https://github.com/webpack-contrib/uglifyjs-webpack-plugin/blob/master/test/all-options.test.js#L70

mattlewis92 commented:
@hulkish done!

michael-ciniawsky commented on src/index.js:
```diff
@@ -1,21 +1,12 @@
-    srcMap = inlineSourceMap.sourcemap;
-  }
+export default function (source) {
+  const userOptions = loaderUtils.getOptions(this) || {};
```
`userOptions` => `options`

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+import MemoryFileSystem from 'memory-fs';
+import webpack from 'webpack';
+
+const istanbulInstrumenterLoader = require.resolve('../src/cjs.js');
```
```diff
- const istanbulInstrumenterLoader
+ const loader
```

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+      path: path.join(__dirname, 'fixtures', 'dist'),
+    },
+    module: {
+      rules: [{
```
```diff
[
  {
    test: /\.js$/
-   loader: istanbulInstrumenterLoader,
+   loader,
-   options: instrumenterOptions,
+   options,
    enforce: 'post'
  }
]
```

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+import path from 'path';
```
Rename to `(webpack|compiler|run).js`

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+  });
+}
+
+export function stripLocalPaths(source) {
```
Move to `test/helpers.js` (Separate file to compiler)

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+}
+
+export function stripLocalPaths(source) {
+  const regexpReplace = new RegExp(__dirname, 'g');
```
`regexReplace` => `replacer || replacee || regex`

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+  }, extra);
+}
+
+export function runWebpack(config) {
```
Make this the `default` export without a name, see below why :)

michael-ciniawsky replied:
```js
export default function (fixture, options, extend = {}) {
  const config = Object.assign({
    ...webpackConfigFactory
   }, extend)
   ...runWebpack
}
```

michael-ciniawsky commented on test/index.test.js:
```diff
@@ -0,0 +1,34 @@
+import { runWebpack, webpackConfigFactory, stripLocalPaths } from './helpers';
```
```js
import webpack from './webpack.js'; // Constistent across the webpack-contrib (`webpack-defaults`)
import { stripLocalPaths } from './helpers'; // Custom helpers per repo
```

michael-ciniawsky commented on test/helpers.js:
```diff
@@ -0,0 +1,40 @@
+
+const istanbulInstrumenterLoader = require.resolve('../src/cjs.js');
+
+export function webpackConfigFactory({ instrumenterOptions, extra, fixtureEntry = 'basic.js' } = {}) {
```
Move into `runWebpack`

michael-ciniawsky commented:
Please also split `chore, test, refactor` tasks into micro PRs accordingly whenever possible, I ensure you it will get reviewed :)

mattlewis92 commented:
@michael-ciniawsky it was kind of tricky to do that, since each subsequent commit depended on changes from the previous commit. It's probably best to just squash and merge these changes into one commit anyway 😄

michael-ciniawsky commented:
Mind to open a PR in `webpack-defaults` with your initial test setup `test/(compiler|run|webpack).js` to start working on [#68 (`webpack-defaukts`)](https://github.com/webpack-contrib/webpack-defaults/issues/68) ? 🙃

michael-ciniawsky commented on package.json:
```diff
@@ -48,7 +48,6 @@
-    "jest-cli": "^20.0.4",
     "lint-staged": "^3.6.1",
     "memory-fs": "^0.4.1",
     "nodemon": "^1.11.0",
```
Not needed anymore since `~webpack-defaults v1.3.0+`

mattlewis92 commented:
Had a minor heart attack for one sec as I messed up the rebase then force pushed the changes and lost all my work... thankfully webstorm has a file restore feature 😅 

@michael-ciniawsky I've addressed your review changes and squashed the commits into 2 - one for upgrading `webpack-defaults` and the other for making the `produceSourceMap` on by default to maintain BC with 2.x + added tests to verify the behaviour. LMK if you'd like me to change anything else 😄

michael-ciniawsky commented on package.json:
```diff
@@ -34,26 +33,29 @@
-    "lint-staged": "^3.4.0",
+    "lint-staged": "^3.6.1",
+    "memory-fs": "^0.4.1",
     "nodemon": "^1.11.0",
```
```diff
- nodemon
```

michael-ciniawsky commented:
Drop the `nodemon` dependency please and g2g :)

mattlewis92 commented:
Done! 💃

mattlewis92 commented:
Awesome thanks @michael-ciniawsky ! Would it be possible for either @d3viant0ne or @bebraw to cut a new beta with this change please? Thanks! 😄 

btw there was actually a fix in with this PR, previously `produceSourceMap` was off and now it's on by default. It's probably worth adding this to the changelog for the release when its generated (although the previous release was broken so I doubt anyone would have had issues with it)

d3viant0ne commented:
@mattlewis92 - Doing that now

d3viant0ne commented:
Done - `3.0.0-beta.1`

mattlewis92 commented:
Ah it's because the webpack build hashes are different, they're non-deterministic by default and not based off the file contents. I will just strip them from the snapshots the same way I do for the file paths, will send a PR later 😀
